bibliopixel/drivers/spi_driver_base.py

```diff
@@ -1,77 +1,168 @@
 from . driver_base import DriverBase
 from .. import log
 
 
-class DriverSPIBase(DriverBase):
-    """
-    Base driver for controling SPI devices on systems like the Raspberry Pi and
-    BeagleBone.
-    """
-    def __init__(self, num, c_order=ChannelOrder.GRB, use_py_spi=True,
-                 dev='/dev/spidev0.0', SPISpeed=2, open=open, gamma=None):
-        super().__init__(num, c_order=c_order, gamma=gamma)
+class SpiBaseInterface(object):
+    """ abstract class for different spi backends"""
 
-        self.dev = dev
-        self.use_py_spi = use_py_spi
-        self._spiSpeed = SPISpeed
-        self._open = open
+    def __init__(self, dev, spi_speed):
+        self._dev = dev
+        self._spi_speed = spi_speed
 
-        # File based SPI requires a bytearray so we have to overwrite _buf
-        if self.use_py_spi:
-            a, b = -1, -1
-            d = self.dev.replace('/dev/spidev', '')
-            s = d.split('.')
-            if len(s) == 2:
-                a = int(s[0])
-                b = int(s[1])
+    def send_packet(self, data):
+        raise NotImplementedError
 
-            if a < 0 or b < 0:
-                error(BAD_FORMAT_ERROR)
+    def compute_packet(self, data):
+        return data
 
-            self._bootstrapSPIDev()
-            import spidev
-            self.spi = spidev.SpiDev()
-            self.spi.open(a, b)
-            self.spi.max_speed_hz = int(self._spiSpeed * 1000000.0)
-            log.info('py-spidev speed @ %.1f MHz',
-                     (float(self.spi.max_speed_hz) / 1000000.0))
-        else:
-            self._buf = bytearray(self._buf)
-            self.spi = self._open(self.dev, 'wb')
-
-    def _bootstrapSPIDev(self):
-        if self.use_py_spi:
-            try:
-                import spidev
-            except:
-                error(CANT_IMPORT_SPIDEV_ERROR)
-
-        if not os.path.exists(self.dev):
+
+# partial source @see https://armbedded.taskit.de/node/318
+# @see <linux/spi/spidev.h>
+# hex numbers from good old strace(1)
+IOC_READ = 0x40000000
+IOC_WRITE = 0x40000000
+
+IOC_BITS_PER_WORD = 0x00016b03
+IOC_MAX_SPEED_HZ = 0x00046b04
+
+
+class SpiFileInterface(SpiBaseInterface):
+    """ using os open/write to send data"""
+
+    def __init__(self, **kwargs):
+        super().__init__(**kwargs)
+        if not os.path.exists(self._dev):
             error(CANT_FIND_ERROR)
 
+        try:
+            import fcntl
+            self._fcntl = fcntl
+        except ImportError:
+            error(CANT_IMPORT_FCNTL_ERROR)
+        self._spi = open(self._dev, 'wb')
+
+        self.speed = self._spi_speed
+        self.bits_per_word = 8
+        log.info('file io spi speed @ %.1f MHz, %d bits per word', self.speed / 1e6, self.bits_per_word)
+
+    def _ioctl(self, command, buffer):
+        self._fcntl.ioctl(self._spi, IOC_WRITE + command, buffer)
+        self._fcntl.ioctl(self._spi, IOC_READ + command, buffer)
+
+    def _set_speed(self, speed):
```

Comment:
Your property/setters are old-style!  New style looks [like this](https://github.com/rec/BiblioPixel/commit/ec7f72b254081368fd973c061e8f0610e68ee0fa).